tests/interfaces/default_qubit_2_integration/test_jax_qnode_integration.py

```diff
@@ -0,0 +1,2143 @@
+class TestQNode:
+    """Test that using the QNode with JAX integrates with the PennyLane
+    stack"""
+
+    def test_execution_with_interface(self, diff_method, grad_on_execution, interface):
+        """Test execution works with the interface"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        @qnode(
+            dev, interface=interface, diff_method=diff_method, grad_on_execution=grad_on_execution
+        )
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(0.2, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        a = np.array(0.1, requires_grad=True)
+        circuit(a)
+
+        assert circuit.interface == interface
+
+        # the tape is able to deduce trainable parameters
+        assert circuit.qtape.trainable_params == [0]
+
+        # gradients should work
+        grad = jax.grad(circuit)(a)
+        assert isinstance(grad, jax.Array)
+        assert grad.shape == ()
+
+    def test_changing_trainability(
+        self, diff_method, grad_on_execution, interface, mocker, tol
+    ):  # pylint:disable=unused-argument
+        """Test changing the trainability of parameters changes the
+        number of differentiation requests made"""
+        if diff_method != "parameter-shift":
+            pytest.skip("Test only supports parameter-shift")
+
+        a = jax.numpy.array(0.1)
+        b = jax.numpy.array(0.2)
+
+        @qnode(dev, interface=interface, diff_method="parameter-shift")
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliY(1)]))
+
+        grad_fn = jax.grad(circuit, argnums=[0, 1])
+        spy = mocker.spy(qml.gradients.param_shift, "transform_fn")
+        res = grad_fn(a, b)
+
+        # the tape has reported both arguments as trainable
+        assert circuit.qtape.trainable_params == [0, 1]
+
+        expected = [-np.sin(a) + np.sin(a) * np.sin(b), -np.cos(a) * np.cos(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # The parameter-shift rule has been called for each argument
+        assert len(spy.spy_return[0]) == 4
+
+        # make the second QNode argument a constant
+        grad_fn = jax.grad(circuit, argnums=0)
+        res = grad_fn(a, b)
+
+        # the tape has reported only the first argument as trainable
+        assert circuit.qtape.trainable_params == [0]
+
+        expected = [-np.sin(a) + np.sin(a) * np.sin(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # The parameter-shift rule has been called only once
+        assert len(spy.spy_return[0]) == 2
+
+        # trainability also updates on evaluation
+        a = np.array(0.54, requires_grad=False)
+        b = np.array(0.8, requires_grad=True)
+        circuit(a, b)
+        assert circuit.qtape.trainable_params == [1]
```

Comment:
So maybe this is just me, but I'm a little confused by why we are setting trainability based on autograd for a jax qnode?

Reply:
hmm that happens in quite a few places. I'd be fine with making it totally jax-like, users shouldn't do this even if we do "support" it.